### PR TITLE
WT-11733 Decouple validation from tracking in the cppsuite

### PR DESCRIPTION
**Pull request by etienneptl**
We should be able to validate a workload without the need for the tracking component to be enabled.
The `validate` function has been updated with a new first argument `tracking_enabled`. This one has to be `true` for the default implementation.

### CODE REVIEW

File: test/cppsuite/src/main/test.cpp [hunk 1]
```diff
@@ -168,11 +168,8 @@ test::run()
         it->finish();
 
     /* Validation stage. */
-    if (_operation_tracker->enabled()) {
-        std::unique_ptr<configuration> tracking_config(_config->get_subconfig(OPERATION_TRACKER));
-        this->validate(_operation_tracker->get_operation_table_name(),
-          _operation_tracker->get_schema_table_name(), _workload_manager->get_database());
-    }
+    this->validate(_operation_tracker->enabled(), _operation_tracker->get_operation_table_name(),
```
Comment (etienneptl) [inline test/cppsuite/src/main/test.cpp:171]: We now always call` validate`. If one overrides this function, they can validate their test without having the tracking component enabled.

File: test/cppsuite/tests/cache_resize.cpp [hunk 2]
```diff
@@ -180,6 +180,7 @@ class cache_resize : public test {
         (void)cache_size_500mb;
 
         /* Open a cursor on the tracking table to read it. */
+        testutil_assert(tracking_enabled);
```
Comment (ajmorton) [inline test/cppsuite/tests/cache_resize.cpp:183]:
```suggestion
        // This validate logic uses data from operation tracking. It must be enabled
        testutil_assert(tracking_enabled);
```

File: test/cppsuite/src/main/test.cpp [hunk 3]
```diff
@@ -168,11 +168,8 @@ test::run()
         it->finish();
 
     /* Validation stage. */
-    if (_operation_tracker->enabled()) {
-        std::unique_ptr<configuration> tracking_config(_config->get_subconfig(OPERATION_TRACKER));
```
Comment (ajmorton) [inline test/cppsuite/src/main/test.cpp:172]: Sidenote, no action required: It seems like we don't have compiler warnings for unused variables, and looking in the cmake/ folder we don't configure any warnings in `COMPILER_DIAGNOSTIC_CXX_FLAGS`.  [WT-8247](https://jira.mongodb.org/browse/WT-8247) is in the defined pipeline to add warnings for cpp code.

**Review by ajmorton — APPROVED** [review]
LGTM 👍 
Could you provide an example where we'd use validation but don't need op tracking to do so?

**Comment by etienneptl** [comment]
> LGTM 👍 Could you provide an example where we'd use validation but don't need op tracking to do so?

@ajmorton, the idea came from WT-11346 where @Sean04 would like to validate his tests without using the tracking component.

File: test/cppsuite/src/main/test.cpp (hunk 3 above)
Comment (etienneptl) [reply test/cppsuite/src/main/test.cpp:172]: fb5c7b9ebe7efe43860c87692292d32d3ec9bd50

File: test/cppsuite/tests/cache_resize.cpp (hunk 2 above)
Comment (etienneptl) [reply test/cppsuite/tests/cache_resize.cpp:186]: fb5c7b9ebe7efe43860c87692292d32d3ec9bd50

File: test/cppsuite/how_to_use_cppsuite.md [hunk 4]
```diff
@@ -103,12 +103,13 @@ In our test `multi_size_inserts` the 10MB updates are too large for cache and no
...
 ```cpp
 void
-validate(const std::string &operation_table_name, const std::string &schema_table_name,
+validate(bool tracking_enabled, const std::string &operation_table_name, const std::string &schema_table_name,
     database &db) override final
 {
     const int64_t ten_mb = 10 * 1024 * 1024;
     const int64_t one_kb = 1024;
 
+    testutil_assert(tracking_enabled);
```
[the remark this answers is not in the document]
Comment (etienneptl) [reply test/cppsuite/how_to_use_cppsuite.md:116]: fb5c7b9ebe7efe43860c87692292d32d3ec9bd50

**Comment by luke-pearson** [comment]
> where @Sean04 would like to validate his tests without using the tracking component.

I am interested in knowing how we will explain the Metrics Monitor which does end of test validation, verses this?

**Comment by ajmorton** [comment]
> > where @Sean04 would like to validate his tests without using the tracking component.
> 
> I am interested in knowing how we will explain the Metrics Monitor which does end of test validation, verses this?

Metrics and validation are independent concepts, no? Validation is checking correctness of the database contents, while metrics monitor is capturing statistics from the test run. As part of collecting metrics we'll check they're in expected ranges, but I'm not aware of any metric checks that would detect invalid keys and the like.

**Comment by luke-pearson** [comment]
> Validation is checking correctness of the database contents, while metrics monitor is capturing statistics from the test run
I think they might still be validating statistical things.

**Comment by etienneptl** [comment]
The annoying thing we noticed with @Sean04 when we tried to use the Metrics monitor is that we have to redefine stats in the cpp suite so we can use them in the config. This is not great in terms of user experience and amount of work. Furthermore, the metrics monitor allows you to do checks vs thresholds. If you want to do more complex checks, you cannot.

**Comment by luke-pearson** [comment]
Can you describe your usage scenario, I'm still reasonably convinced this is something a metrics monitor would handle.

**Comment by etienneptl** [comment]
A sample validation would be to check stats. However, to be able to check stats, you need to re-define then and add some logic to the doce in multiple places, see `cc_pages_removed` as an example. It does not seem efficient, we cannot do this for every single stat.

**Comment by luke-pearson** [comment]
So you want the test to be able to define custom statistic validation logic, do tests overload validate()? I forget. It feels like the issue we have is that statistics are too baked into the config and that makes customizing them hard. I don't feel like decoupling validation from tracking is necessarily the fix then. If it works it works, but perhaps theres a better way of doing this?

**Comment by etienneptl** [comment]
Yes, tests can overload the `validate` function. I still believe calling `validate` should not depend on the `tracking` component being enabled. I see the tracking component as something that can gather additional data that can be used for validation.

> It feels like the issue we have is that statistics are too baked into the config and that makes customizing them hard

Yes exactly!

> but perhaps theres a better way of doing this?

We definitely need to improve how we can check for stats without adding code. We also need anyone to have access to the validate function without the need of enabling other stuff. Decoupling is always better imo.

**Review by luke-pearson — APPROVED** [review]
LGTM.

**Comment by etienneptl** [comment]
@luke-pearson, I created [WT-11742](https://jira.mongodb.org/browse/WT-11742) to improve the Metrics Monitor in the cppsuite.

**Comment by etienneptl** [comment]
evergreen merge

**Pull request merged**